sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -0,0 +1,150 @@
+ */
+object NestedColumnAliasing {
+
+  def unapply(plan: LogicalPlan)
+    : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = plan match {
+    case Project(_, child) if canProjectPushThrough(child) =>
+      getAliasSubMap(plan, child)
+    case _ => None
+  }
+
+  /**
+   * Replace nested columns to prune unused nested columns later.
+   */
+  def replaceToAliases(
+      plan: LogicalPlan,
+      nestedFieldToAlias: Map[GetStructField, Alias],
+      attrToAliases: Map[ExprId, Seq[Alias]]): LogicalPlan = plan match {
+    case Project(projectList, child) =>
+      Project(
+        getNewProjectList(projectList, nestedFieldToAlias),
+        replaceChildrenWithAliases(child, attrToAliases))
+  }
+
+  /**
+   * Return a replaced project list.
+   */
+  private def getNewProjectList(
+      projectList: Seq[NamedExpression],
+      nestedFieldToAlias: Map[GetStructField, Alias]): Seq[NamedExpression] = {
+    projectList.map(_.transform {
+      case f: GetStructField if nestedFieldToAlias.contains(f) =>
+        nestedFieldToAlias(f).toAttribute
+    }.asInstanceOf[NamedExpression])
+  }
+
+  /**
+   * Return a plan with new childen replaced with aliases.
+   */
+  private def replaceChildrenWithAliases(
+      plan: LogicalPlan,
+      attrToAliases: Map[ExprId, Seq[Alias]]): LogicalPlan = {
+    plan.withNewChildren(plan.children.map { plan =>
+      Project(plan.output.flatMap(a => attrToAliases.getOrElse(a.exprId, Seq(a))), plan)
+    })
+  }
+
+  /**
+   * Returns true for those operators that project can be pushed through.
+   */
+  private def canProjectPushThrough(plan: LogicalPlan) = plan match {
+    case _: GlobalLimit => true
+    case _: LocalLimit => true
+    case _: Repartition => true
+    case _: Sample => true
+    case _ => false
+  }
+
+  /**
+   * Return root references that are individually accessed as a whole, and `GetStructField`s.
+   */
+  private def collectRootReferenceAndGetStructField(plan: LogicalPlan): Seq[Expression] = {
+    def helper(e: Expression): Seq[Expression] = e match {
+      case _: AttributeReference | _: GetStructField => Seq(e)
+      case es if es.children.nonEmpty => es.children.flatMap(helper)
+      case _ => Seq.empty
+    }
+    plan.expressions.flatMap(helper)
+  }
+
+  /**
+   * Return two maps in order to replace nested fields to aliases.
+   *
+   * 1. GetStructField -> Alias: A new alias is created for each nested field.
+   * 2. ExprId -> Seq[Alias]: A reference attribute has multiple aliases pointing it.
+   */
+  private def getAliasSubMap(plans: LogicalPlan*)
+    : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = {
+    val (nestedFieldReferences, otherRootReferences) = plans
+      .map(collectRootReferenceAndGetStructField).reduce(_ ++ _).partition {
+        case _: GetStructField => true
+        case _ => false
+      }
+
+    val aliasSub = nestedFieldReferences.asInstanceOf[Seq[GetStructField]]
```

Comment:
nit: Drop `.asInstanceOf[Seq[GetStructField]]`.

Reply:
Thanks.

Reply:
Oops. I forgot that I added that for Scala issue. We need this here for the below `flatMap`.
```
[error] [warn] /Users/dongjoon/PRS/SPARK-26975-ALIAS/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:117: non-variable type argument org.apache.spark.sql.catalyst.expressions.GetStructField in type pattern Seq[org.apache.spark.sql.catalyst.expressions.GetStructField] (the underlying of Seq[org.apache.spark.sql.catalyst.expressions.GetStructField]) is unchecked since it is eliminated by erasure
[error] [warn]       .flatMap { case (attr, nestedFields: Seq[GetStructField]) =>
[error] [warn] 
```

I'll keep the current one.

Reply:
ok

Reply:
Thanks!